include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce_then_scan.h

```diff
@@ -604,109 +610,95 @@ struct __parallel_reduce_then_scan_scan_submitter<__sub_group_size, __max_inputs
-                if constexpr (!__is_inclusive)
-                {
-                    auto __global_id = __ndi.get_global_linear_id();
-                    if (__global_id == __num_work_items - 1)
-                    {
-                        std::size_t __last_idx_in_block = std::min(__n - 1, __max_block_size * (__block_num + 1) - 1);
-                        __tmp_storage[__num_sub_groups_global] = __in_rng[__last_idx_in_block];
+                        __sub_group_carry.__setup(__tmp_ptr[__num_sub_groups_global]);
                     }
                 }
 
                 // step 5) apply global carries
-                size_t __subgroup_start_idx = __group_start_idx + (__sub_group_id * __inputs_per_sub_group);
-                size_t __start_idx = __subgroup_start_idx + __sub_group_local_id;
+                std::size_t __subgroup_start_idx = __group_start_idx + (__sub_group_id * __inputs_per_sub_group);
+                std::size_t __start_idx = __subgroup_start_idx + __sub_group_local_id;
 
                 if (__sub_group_carry_initialized)
                 {
                     __scan_through_elements_helper<__sub_group_size, __is_inclusive,
                                                    /*__init_present=*/true,
                                                    /*__capture_output=*/true, __max_inputs_per_item>(
-                        __sub_group, __unary_op, __binary_op, __sub_group_carry, __in_rng, __out_rng, __start_idx, __n,
-                        __inputs_per_item, __subgroup_start_idx, __sub_group_id, __active_subgroups);
-
-                    __sub_group_carry.__destroy();
+                        __sub_group, __gen_scan_input, __scan_pred, __reduce_op, __write_op, __sub_group_carry,
+                        __in_rng, __out_rng, __start_idx, __n, __inputs_per_item, __subgroup_start_idx, __sub_group_id,
+                        __active_subgroups);
                 }
                 else // first group first block, no subgroup carry
                 {
                     __scan_through_elements_helper<__sub_group_size, __is_inclusive,
                                                    /*__init_present=*/false,
                                                    /*__capture_output=*/true, __max_inputs_per_item>(
-                        __sub_group, __unary_op, __binary_op, __sub_group_carry, __in_rng, __out_rng, __start_idx, __n,
-                        __inputs_per_item, __subgroup_start_idx, __sub_group_id, __active_subgroups);
+                        __sub_group, __gen_scan_input, __scan_pred, __reduce_op, __write_op, __sub_group_carry,
+                        __in_rng, __out_rng, __start_idx, __n, __inputs_per_item, __subgroup_start_idx, __sub_group_id,
+                        __active_subgroups);
+                }
+                //if at the last element in the sequence, then we need to write out the last carry out
+                if (__sub_group_local_id == 0 && (__active_groups == __g + 1) &&
```

Comment:
Not asking to change, but just pointing out that these conditions with `+ 1` tripped me up since I'm so used to seeing conditions like this formatted as `g == __active_groups - 1` instead. Maybe we can consider switching the condition around to be more consistent with the rest of the codebase, but it's not a big deal.

Reply:
Thats fair.  I did this with the thought that `__active_[sub]groups` could be 0, and both are unsigned types.  I think now in the current state of the codebase they cannot be 0, but I was trying to avoid underflow issues.

Its also reasonable to make these signed types.  They wont be overflowing their respective types if we make them signed.

Reply:
I improved the comment to make this more clear (hopefully)